Text results for contributors

janno42 opened the pull request:
fix #551 

This allows staff users and editors to define on the course edit page, which contributors can see which comments after the course was published.

karyon commented on evap/staff/forms.py:
```diff
@@ -204,6 +213,14 @@ def validate_unique(self):
+        self.instance.responsible = responsibility == Contribution.IS_RESPONSIBLE
+        self.instance.can_edit = (responsibility == Contribution.IS_RESPONSIBLE) or (responsibility == Contribution.IS_EDITOR)
+        self.instance.save()
+        return self.instance
```
please do the stuff first and then do `return super().save(*args, **kwargs)` without `self.instance.save()`, i have the feeling that right now you would change values in the db even if the formset generates some errors.

karyon commented:
* why is it no longer possible to make a contributor the responsible by clicking on their "responsible"-button? right now i have to de-select the current responsible and then change the new one.
* when loading the page, the responsible has "own" comments selected. making him editor and then responsible again fixes that. i think responsibility should always imply "all" availability, even in the database. the form's save() method should enforce that, and the migration would need to handle that as well.
* when clicking those buttons, the tooltip stays there. that was ok for the questionnaires, but it looks really bad for the buttons imo, especially because a tooltip on the lower buttons overlaps the upper buttons
* just an idea, i don't know whether it makes sense: maybe editor should imply "course" availability?

janno42 commented:
1. will change
2. i forgot that in the migration, yes. i will also add it in save.
3. no idea what to do
4. nah, i don't think that should have anything to do with that.

janno42 commented:
addressed 2 and 3. 1 and 4 are okay.
there are still webtest form parsing errors left.

karyon commented on evap/contributor/templates/contributor_course_form.html:
```diff
@@ -28,43 +28,60 @@
+                                        <label class="btn btn-sm btn-option{% if form.responsibility.value == "RESPONSIBLE" %} disabled{% endif %}{% if form.comment_visibility.value == choice.choice_value %} active{% endif %}{% if form.comment_visibility.errors %} choice-error{% endif %}" name="{{ choice.name }}" data-toggle="tooltip" data-placement="top" title=
+                                        {% if choice.choice_value == "OWN" %}"{% trans "Default value. Will only see own comments." %}"
+                                        {% elif choice.choice_value == "COURSE" %}"{% trans "Will see own comments and comments about the course." %}"
+                                        {% elif choice.choice_value == "ALL" %}"{% trans "Will see all comments about the course and about all contributors." %}"
```
please put these if elifs into a tag, or put the whole widget into a separate file. i'd prefer the latter to duplicate less code with staff_course_form.html

karyon replied:
also, please use more line breaks in the label tag

karyon commented:
i have a feeling i commented this already, but couldn't find it. do you think it makes sense to show the always disabled responsible-button on the contributor pages? it think you should at least add some explanation why that can't be changed, if not design the ui controls differently for contributors.

karyon commented on evap/results/views.py:
```diff
@@ -134,6 +134,11 @@ def user_can_see_text_answer(user, text_answer, public_view=False):
+        if text_answer.contribution.course.contributions.filter(contributor=user, comment_visibility=Contribution.ALL_COMMENTS).exists():
+            return True
+        if text_answer.contribution.is_general and \
+            text_answer.contribution.course.contributions.filter(contributor=user, comment_visibility=Contribution.COURSE_COMMENTS).exists():
```
what about delegates?

good luck :)

karyon commented on evap/staff/forms.py:
```diff
@@ -183,16 +183,25 @@ def validate_unique(self):
+        elif self.instance.can_edit:
+            self.initial['responsibility'] = Contribution.IS_EDITOR
+        else:
+            self.initial['responsibility'] = Contribution.IS_CONTRIBUTOR
```
if this works, could you please check the difference between self.initial[fieldname] and self.fields[fieldname].initial?

janno42 replied:
From the documentation: `Note that if a Field defines initial and you include initial when instantiating the Form, then the latter initial will have precedence.`
Otherwise they do the same thing.

karyon replied:
alright, if you're not using those in semantically different ways, please be consistent and use only one way.

janno42 replied:
check

karyon commented on evap/evaluation/templates/evap_course_edit_js.html:
```diff
@@ -2,48 +2,90 @@
+                // disable all other responsible buttons
+                for(var j = 0; j < contributorCount; j++){
+                    if (j != i)
+                        disableButtons("input[name=contributions-" + j + "-responsibility][value=RESPONSIBLE]");
```
this shouldn't be necessary right?

janno42 replied:
when selecting a responsible button, all other responsible buttons must be disabled, because it should only be possible to select one contributor as responsible.

karyon replied:
yeah i know... for some reason i thought that shouldn't be necessary here, but i don't remember why...

karyon commented on evap/evaluation/templates/evap_course_edit_js.html:
```diff
@@ -2,48 +2,90 @@
-                    // enable can-edit
-                    $("#id_contributions-" + i + "-can_edit").prop("disabled", false);
+            }
+            // unchecked responsible right now
```
could you clarify that comment? this is an if-elseif, a "right now" in between doesn't make sense to me

janno42 replied:
the part below is done after a responsible button has been unchecked. what should i write instead?

karyon replied:
"user unchecked responsible button" or "if the responsible button is deselected" or "user changed selection from responsible to something else". at first i misunderstood this as a description of the state, which was confusing in between the if-elseif, but it should be more a description of the action that was performed.

janno42 replied:
check

karyon commented on evap/evaluation/templates/evap_course_edit_js.html:
```diff
@@ -2,48 +2,90 @@
+            // don't do anything on disabled buttons
+            if ($(this).hasClass("disabled"))
+                return false;
         }
```
what is this good for?

janno42 replied:
when not using this, buttons with the class "disabled" can still be toggled.

karyon replied:
could you state that problem somehow in the comment?

janno42 replied:
check

karyon commented on evap/evaluation/templates/evap_course_edit_js.html:
```diff
@@ -2,48 +2,90 @@
+    }
+
+    function assignClickHandlers() {
+        for(var i = 0; i < getContributorCount(); i++)
```
please put the result of getContributorCount() into a variable

janno42 replied:
check

karyon commented on evap/evaluation/templates/evap_course_edit_js.html:
```diff
@@ -2,48 +2,90 @@
-            if($("#id_contributions-" + i + "-responsible").is(":checked")) {
-                $("#id_contributions-" + i + "-can_edit").prop("checked", true);
-                $("#id_contributions-" + i + "-can_edit").prop("disabled", true);
+            if($("input[name=contributions-" + i + "-responsibility][value=RESPONSIBLE]").is(":checked")) {
```
you use `.is(":checked")` here but `.prop("checked")` above, please make it consistent

janno42 replied:
check

karyon commented on evap/evaluation/templates/evap_course_edit_js.html:
```diff
@@ -2,48 +2,90 @@
+                for(var j = 0; j < getContributorCount(); j++){
+                    if (j != i)
+                        disableButtons("input[name=contributions-" + j + "-responsibility][value=RESPONSIBLE]");
+                }
```
where do you initially check "all" in the comment widget for responsibles? why doesn't that happen here?

janno42 replied:
responsible contributors already have "all" saved in the db. should i make a second override here?

karyon replied:
no, probably not.

karyon commented on evap/evaluation/templates/evap_course_edit_js.html:
```diff
@@ -2,48 +2,90 @@
-                $("#id_contributions-" + i + "-can_edit").prop("disabled", true);
+            if($("input[name=contributions-" + i + "-responsibility][value=RESPONSIBLE]").is(":checked")) {
+                // disable comment buttons for responsible contributor
+                disableButtons("input[name=contributions-" + i + "-comment_visibility]");
```
in the UI this is called "comment availability", here its visibility, please make it consistent

janno42 replied:
check

janno42 commented:
addressed comments. now also delegates of qualified users can see comments and the responsible button is not shown for contributors (except when active).

karyon commented on evap/evaluation/templates/comment_visibility_buttons.html:
```diff
@@ -0,0 +1,19 @@
+        {% if form.responsibility.value == "RESPONSIBLE" %}
+            {% include_choice_button form.comment_visibility choice False tooltip %}
+        {% else %}
+            {% include_choice_button form.comment_visibility choice True tooltip %}
```
can you do something like `form.responsibility.value == "RESPONSIBLE" as disabled` and then use `disabled` as parameter? (i mixed up True and False but you get the idea)

janno42 replied:
i think variable assignment is only possible with `with`and something like `{% with disabled = form.responsibility.value == "RESPONSIBLE" %}` does not work. if you have any other idea, please let me know. and keep in mind that we would actually need `enabled`, and as you know django template doesn't like negations.

karyon commented on evap/evaluation/templates/responsibility_buttons.html:
```diff
@@ -0,0 +1,27 @@
+            {% if form.responsibility.value == "RESPONSIBLE" and can_change_responsible|is_false %}
+                {% include_choice_button form.responsibility choice False tooltip %}
+            {% else %}
+                {% include_choice_button form.responsibility choice True tooltip %}
```
same here

karyon commented on evap/results/views.py:
```diff
@@ -135,5 +135,12 @@ def user_can_see_text_answer(user, text_answer, public_view=False):
             return True
         if text_answer.contribution.course.is_user_responsible_or_delegate(user):
             return True
```
these two lines should be superfluous now if i got that correctly. also, please put `list(user.represented_users.all()) into a variable to avoid hitting the database twice in the common case (which is (i think) that a comment is not visible)

karyon commented on evap/staff/forms.py:
```diff
@@ -204,6 +213,14 @@ def validate_unique(self):
+        self.instance.responsible = responsibility == Contribution.IS_RESPONSIBLE
+        self.instance.can_edit = (responsibility == Contribution.IS_RESPONSIBLE) or (responsibility == Contribution.IS_EDITOR)
+        if responsibility == Contribution.IS_RESPONSIBLE:
+            self.instance.comment_visibility = Contribution.ALL_COMMENTS
```
please put `responsibility == Contribution.IS_RESPONSIBLE` and `responsibility == Contribution.IS_EDITOR` into a variable

karyon commented:
done reviewing.

karyon commented:
:+1:
